aiosmtpd/proxy_protocol.py

```diff
@@ -0,0 +1,551 @@
+        0x25: "SSL_KEY_ALG",
+        0x30: "NETNS",
+    }
+
+    def __init__(self, *args, _tlv_loc: Dict[str, int], **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tlv_loc = _tlv_loc
+
+    def __getattr__(self, item):
+        return self.get(item)
+
+    def __eq__(self, other):
+        return super().__eq__(other)
+
+    def same_attribs(self, _raises: bool = False, **kwargs) -> bool:
+        """
+        Helper function to test whether attribute(s) exists, and whether the attribute's
+        value is as expected. For ProxyTLV, since it inherits from dict, you can also
+        use direct dict comparison.
+
+        :param _raises: If True, raises an Error instead of returning bool
+        """
+        for k, v in kwargs.items():
+            actual = self.get(k, _NOT_FOUND)
+            if actual is _NOT_FOUND:
+                if _raises:
+                    raise KeyError(f"notfound:{k}")
+                else:
+                    return False
+            if actual != v:
+                if _raises:
+                    raise ValueError(f"mismatch:{k} actual={actual!r} expect={v!r}")
+                else:
+                    return False
+        return True
+
+    @classmethod
+    def parse(
+        cls,
+        data: Union[bytes, bytearray],
+        partial_ok: bool = True,
+        strict: bool = False,
+    ) -> Tuple[Dict[str, Any], Dict[str, int]]:
+        """
+        Parse a bunch of bytes into TLV Vectors.
+
+        :param data: The bunch of bytes to parse
+        :param partial_ok: Keep result of parsing so far if error encountered
+        :param strict: If true, reject unrecognized TYPEs
+        """
+        rslt: Dict[str, Any] = {}
+        tlv_loc: Dict[str, int] = {}
+
+        def _pars(chunk: Union[bytes, bytearray], *, offset: int):
+            i = 0
+            while i < len(chunk):
+                typ = chunk[i]
+                len_ = int.from_bytes(chunk[i + 1 : i + 3], "big")
+                val = chunk[i + 3 : i + 3 + len_]
+                if len(val) < len_:
+                    raise MalformedTLV(f"TLV 0x{typ:02X} is malformed!")
+                typ_name = cls.PP2_TYPENAME.get(typ)
+                if typ_name is None:
+                    typ_name = f"x{typ:02X}"
+                    if strict:
+                        raise UnknownTypeTLV(typ_name)
+                tlv_loc[typ_name] = offset + i
+                if typ_name == "SSL":
+                    rslt["SSL_CLIENT"] = val[0]
+                    rslt["SSL_VERIFY"] = int.from_bytes(val[1:5], "big")
+                    try:
+                        _pars(val[5:], offset=i)
+                        rslt["SSL"] = True
+                    except MalformedTLV:
+                        rslt["SSL"] = False
+                        if not partial_ok:
+                            raise
+                        else:
+                            return
+                else:
+                    rslt[typ_name] = val
```

Comment:
`val` is a `bytearray` here. It might be my lack of Python expertise, but I think that a `bytes` object might be more useful here, because one usually does not need to handle individual characters, but instead process e.g. the `UNIQUE_ID` as-is.

Reply:
`bytes` and `bytearray` are practically the same in Python. They are comparable, slice-able, decode-able, etc.

The only difference is that the latter is mutable, while the former is not.

Reply:
Understood, thank you for the clarification. Then this is good and I'll have to recheck what I did wrong in my test :smile:

Reply:
Just remember you can't directly compare `bytes` to `str`. So the following
    b"123" == "123"
will result in an error, as `str` is "Unicode codepoints". You need to either `decode` the bytes, or `encode` the str, like so:
    (b"123").decode() == "123"
    b"123" == "123".encode()
or if comparing a variable to a literal, don't forget the `b` prefix, like so:
    proxy_data.tlv.UNIQUE_ID = b"some_unique_id"